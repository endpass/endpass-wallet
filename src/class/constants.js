export const REQUEST_TIMEOUT_MSEC = 30000;

export default {
  REQUEST_TIMEOUT_MSEC,
};
