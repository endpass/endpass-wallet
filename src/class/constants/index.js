export { INPAGE_EVENT, INPAGE_ID_PREFIX } from './InpageProvider';

export const REQUEST_TIMEOUT_MSEC = 30000;

export const STORAGE_USER_META_KEY = 'endpass-user-meta';

export const AVAILABLE_USER_META_PROPS = ['activeAccount'];

export const PROXY_REQUEST_PREFIX = 'endpass-identity';

export const HARDWARE_DERIVIATION_PATH = `m/44'/60'/0'/0/`;

export default {
  REQUEST_TIMEOUT_MSEC,
  STORAGE_USER_META_KEY,
  AVAILABLE_USER_META_PROPS,
  PROXY_REQUEST_PREFIX,
  HARDWARE_DERIVIATION_PATH,
};
