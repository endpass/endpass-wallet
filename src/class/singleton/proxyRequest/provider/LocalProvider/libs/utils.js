import regPath from 'path-to-regexp';

const utils = {
  getRouteKey(routesRegexp, routePath) {
    return Object.keys(routesRegexp).find(key => {
      const re = routesRegexp[key];
      return re.test(routePath);
    });
  },
  getPath(serverUrl, url) {
    const levels = url.split(serverUrl);
    const keysPath = levels.pop() || '';
    const ret = keysPath[0] === '/' ? keysPath.substring(1) : keysPath;

    return ret;
  },
  getPathArgs(routesRegexp, key, routePath) {
    const keys = regPath.parse(key);
    const re = routesRegexp[key];
    re.lastIndex = 0; // drop exec position

    const values = re.exec(routePath);
    return keys.reduce((map, item, index) => {
      if (item && item.name) {
        map[item.name] = values[index];
      }
      return map;
    }, {});
  },
};

export default utils;
