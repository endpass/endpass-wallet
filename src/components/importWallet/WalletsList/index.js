export { default } from './WalletsList.vue';
