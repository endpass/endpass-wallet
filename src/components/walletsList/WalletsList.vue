<template>
  <div class="column">
    <div
      v-if="!isLoading && !addresses.length"
      class="buttons"
    >
      <v-button
        :loading="isImporting"
        class-name="is-primary is-medium"
        @click="getNextAddressess"
      >
        Load wallets
      </v-button>
    </div>
    <v-spinner
      v-if="isLoading"
      :is-loading="true"
      label="Please, allow access for import"
      is-label-under-spinner
    />
    <div v-show="!isLoading && addresses.length">
      <wallet-item
        v-for="address in addresses"
        :key="address"
        :address="address"
        :is-active="address === activeAddress"
        @click="setActiveAddress(address)"
      />
      <v-pagination @input="changePage" />
      <div class="buttons">
        <v-button
          :disabled="!activeAddress"
          :loading="isImporting"
          class-name="is-primary is-medium"
          data-test="import-wallet-button"
          @click="handleImport"
        >
          Import
        </v-button>
      </div>
    </div>
    <password-modal
      v-if="isPasswordModal"
      @close="togglePasswordModal"
      @confirm="handleImportByPassword"
    />
  </div>
</template>

<script>
import { mapActions, mapGetters } from 'vuex';
import modalMixin from '@/mixins/modal';
import PasswordModal from '@/components/modal/PasswordModal';
import WalletItem from './WalletItem.vue';

export default {
  name: 'WalletsList',
  props: {
    type: {
      type: String,
      required: true,
    },
    autoLoad: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    activeAddress: null,
    isLoading: false,
    isImporting: false,
    isPasswordModal: false,
    addresses: [],
    offset: 0,
    limit: 10,
  }),
  computed: {
    ...mapGetters('accounts', ['isHDv3WalletByType']),
  },
  methods: {
    ...mapActions('accounts', [
      'addPublicWallet',
      'getNextWalletsFromHd',
      'addHdChildWallets',
    ]),

    async getNextAddressess({ offset = this.offset, limit = this.limit } = {}) {
      this.isLoading = true;

      try {
        this.addresses = await this.getNextWalletsFromHd({
          offset,
          limit,
          walletType: this.type,
        });
      } catch (error) {
        this.$notify(error);
      } finally {
        this.isLoading = false;
      }
    },
    async changePage(page) {
      this.offset = this.limit * (page - 1);
      await this.getNextAddressess();
    },
    async handleImportByPassword(password) {
      await this.addWallet(password);
    },
    async handleImport() {
      if (this.isHDv3WalletByType(this.type)) {
        this.isPasswordModal = true;
      } else {
        await this.addWallet();
      }
    },
    async addWallet(password) {
      this.isImporting = true;
      try {
        const { activeAddress: address, type } = this;
        const index = this.addresses.indexOf(address);
        if (index === -1) {
          throw new Error('Address not selected');
        }

        if (this.isHDv3WalletByType(type)) {
          await this.addHdChildWallets({
            address,
            index,
            password,
            type,
          });
        } else {
          const info = { type, index };
          await this.addPublicWallet({ address, info });
        }
        this.$router.push('/');
      } catch (e) {
        this.$notify({
          title: 'Importing error',
          text: 'An error occurred while importing wallet. Please try again.',
          type: 'is-danger',
        });
      } finally {
        this.isImporting = false;
      }
    },
    setActiveAddress(address) {
      this.activeAddress = address;
      this.$emit('select', address);
    },
  },
  mounted() {
    if (this.autoLoad) {
      this.getNextAddressess();
    }
  },
  mixins: [modalMixin],
  components: {
    WalletItem,
    PasswordModal,
  },
};
</script>

<style lang="scss"></style>
