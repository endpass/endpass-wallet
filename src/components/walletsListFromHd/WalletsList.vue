<template>
  <div class="column">
    <div
      v-if="!isGettingAddresses && isEmptyAddresses"
      class="buttons"
    >
      <v-button
        :loading="isImporting"
        class-name="is-primary is-medium"
        @click="getNextAddressess"
      >
        Load wallets
      </v-button>
    </div>
    <v-spinner
      v-if="isGettingAddresses"
      :is-loading="true"
      label="Please, allow access for import"
      is-label-under-spinner
    />
    <div v-show="!isGettingAddresses && !isEmptyAddresses">
      <wallet-item
        v-for="(address, index) in addresses"
        :key="address"
        :address="address"
        :is-used="isUsedWallet(address)"
        :is-active="address === selectedAddress"
        @click="updateSelectedAddress(address, index)"
      />
      <v-pagination @input="changePage" />
      <slot name="buttons" />
    </div>
  </div>
</template>

<script>
import { mapActions, mapState } from 'vuex';
import WalletItem from './WalletItem.vue';

export default {
  name: 'WalletsList',
  props: {
    type: {
      type: String,
      required: true,
    },
    autoLoad: {
      type: Boolean,
      default: false,
    },
    isImporting: {
      type: Boolean,
      default: false,
    },
    isGettingAddresses: {
      type: Boolean,
      default: false,
    },
  },
  data: () => ({
    offset: 0,
    limit: 10,
    addresses: [],
    selectedAddress: null,
  }),
  computed: {
    isEmptyAddresses() {
      return this.addresses.length === 0;
    },
    ...mapState('accounts', ['wallets']),
  },
  methods: {
    ...mapActions('accounts', ['getNextWalletsFromHd']),

    async getNextAddressess({ offset = this.offset, limit = this.limit } = {}) {
      this.$emit('input', true);
      const { type: walletType } = this;
      try {
        this.addresses = await this.getNextWalletsFromHd({
          offset,
          limit,
          walletType,
        });
      } catch (error) {
        this.$notify(error);
      } finally {
        this.$emit('input', false);
      }
    },

    async changePage(page) {
      this.offset = this.limit * (page - 1);
      await this.getNextAddressess();
    },

    /**
     * @param {string} address
     * @param {number} index
     */
    updateSelectedAddress(address, index) {
      this.selectedAddress = address;
      this.$emit('select', { address, index: index + this.offset });
    },

    /**
     * @param {string} address
     * @returns {boolean}
     */
    isUsedWallet(address) {
      const account = this.wallets[address];
      return account ? !account.isPublic && !account.isHardware : false;
    },
  },
  mounted() {
    if (this.autoLoad) {
      this.getNextAddressess();
    }
  },
  components: { WalletItem },
  model: {
    prop: 'isGettingAddresses',
    event: 'input',
  },
};
</script>
