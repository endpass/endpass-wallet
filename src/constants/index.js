export { INPAGE_EVENT, INPAGE_ID_PREFIX } from './InpageProvider';

export const MAIN_NET_ID = 1;

export const NETWORK_URL = Object.freeze({
  ETH: [
    'wss://eth-mainnet.endpass.com:2084',
    'wss://eth-mainnet.endpass.com',
    'https://eth-mainnet.endpass.com:2083',
    `https://mainnet.infura.io/${ENV.infuraConf.key}`,
  ],
  ROP: [
    'wss://eth-ropsten.endpass.com:2084',
    'wss://eth-ropsten.endpass.com',
    'https://eth-ropsten.endpass.com:2083',
    `https://ropsten.infura.io/${ENV.infuraConf.key}`,
  ],
  RIN: `https://rinkeby.infura.io/${ENV.infuraConf.key}`,
  ETC: [
    'wss://etc-mainnet.endpass.com:2084',
    'wss://etc-mainnet.endpass.com',
    'https://etc-mainnet.endpass.com:2083',
    'https://etc-geth.0xinfra.com',
  ],
});

export const DEFAULT_NETWORKS = Object.freeze([
  {
    id: 1,
    networkType: 'main',
    currency: 1,
    name: 'Main',
    url: NETWORK_URL.ETH,
  },
  {
    id: 3,
    name: 'Ropsten',
    networkType: 'ropsten',
    currency: 2,
    url: NETWORK_URL.ROP,
  },
  {
    id: 4,
    name: 'Rinkeby',
    networkType: 'rinkeby',
    currency: 2,
    url: NETWORK_URL.RIN,
  },
  {
    id: 61,
    name: 'Ethereum classic',
    networkType: 'ethClassic',
    currency: 3,
    url: NETWORK_URL.ETC,
  },
]);

export const CURRENCIES = Object.freeze([
  {
    name: 'ETH',
    id: 1,
  },
  {
    name: 'ETH-TEST',
    id: 2,
  },
  {
    name: 'ETC',
    id: 3,
  },
]);

export const AVAILABLE_FIAT_CURRENCIES = Object.freeze(
  // prettier-ignore
  ['USD', 'AUD', 'BRL', 'CAD', 'CHF', 'CLP', 'CNY', 'CZK', 'DKK', 'EUR', 'GBP', 'HKD', 'HUF', 'IDR', 'ILS', 'INR', 'JPY', 'KRW', 'MXN', 'MYR', 'NOK', 'NZD', 'PHP', 'PKR', 'PLN', 'RUB', 'SEK', 'SGD', 'THB', 'TRY', 'TWD', 'ZAR'],
);

export const IDENTITY_MODE = Object.freeze({
  DEFAULT: 'default',
  CUSTOM: 'custom',
  LOCAL: 'local',
});

export const PROXY_REQUEST_PREFIX = 'endpass-identity';

export const STORAGE_USER_META_KEY = 'endpass-user-meta';

export const AVAILABLE_USER_META_PROPS = ['activeAccount'];

export const REQUEST_TIMEOUT_MSEC = 30000;

export const HARDWARE_DERIVIATION_PATH = `m/44'/60'/0'/0/`;

export const HARDWARE_WALLET_TYPE = Object.freeze({
  TREZOR: 'TrezorAccount',
  LEDGER: 'LedgerAccount',
});

export const WALLET_TYPE = Object.freeze({
  PUBLIC: 'PublicAccount',
  HD_PUBLIC: 'HDPublicAccount',
  HD_MAIN: 'HDMainAccount',
  ...HARDWARE_WALLET_TYPE,
});

export const DAPP_WHITELISTED_METHODS = [
  'personal_sign',
  'eth_personalSign',
  'eth_signTypedData',
  'eth_sendTransaction',
];

export default {
  AVAILABLE_FIAT_CURRENCIES,
  MAIN_NET_ID,
  DEFAULT_NETWORKS,
  CURRENCIES,
  IDENTITY_MODE,
  PROXY_REQUEST_PREFIX,
  STORAGE_USER_META_KEY,
  AVAILABLE_USER_META_PROPS,
  REQUEST_TIMEOUT_MSEC,
  HARDWARE_DERIVIATION_PATH,
  HARDWARE_WALLET_TYPE,
  WALLET_TYPE,
  DAPP_WHITELISTED_METHODS,
};
