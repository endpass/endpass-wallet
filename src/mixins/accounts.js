// Mixin for common methods for Ethereum accounts
import web3 from 'web3';

const { fromWei } = web3.utils;

export default {
  computed: {
    address() {
      return this.$store.state.accounts.address;
    },

    balance() {
      return this.$store.state.accounts.balance === null
        ? null
        : fromWei(this.$store.getters['accounts/balance']);
    },
  },
};
