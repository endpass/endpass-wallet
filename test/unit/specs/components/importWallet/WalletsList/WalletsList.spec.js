import { shallow, createLocalVue } from '@vue/test-utils';
import Notifications from 'vue-notification';
import Vuex from 'vuex';
import VeeValidate from 'vee-validate';
import VueRouter from 'vue-router';

import { testUtils } from '@endpass/utils';

import WalletsList from '@/components/importWallet/WalletsList/WalletsList';
import { WALLET_TYPE } from '@/constants';

import { addresses, address } from 'fixtures/accounts';

const addressPos = addresses.indexOf(address);

const localVue = createLocalVue();

localVue.use(Vuex);
localVue.use(VueRouter);
localVue.use(VeeValidate);
localVue.use(Notifications);

jest.useFakeTimers();

describe('WalletsList', () => {
  let wrapper;
  let actions;
  let router;

  beforeEach(() => {
    actions = {
      addPublicWallet: jest.fn(),
      getNextWalletsFromHd: jest.fn(),
      addHdChildWallets: jest.fn(),
    };
    const storeOptions = {
      modules: {
        accounts: {
          namespaced: true,
          actions,
          getters: {
            isHDv3WalletByType: jest.fn(),
          },
        },
      },
    };
    router = new VueRouter();
    const store = new Vuex.Store(storeOptions);
    wrapper = shallow(WalletsList, {
      localVue,
      store,
      router,
      stubs: testUtils.generateStubs(WalletsList),
      propsData: {
        type: WALLET_TYPE.HD_PUBLIC,
      },
    });
  });

  describe('render', () => {
    it('should be a Vue component', () => {
      expect(wrapper.name()).toBe('WalletsList');
      expect(wrapper.isVueInstance()).toBeTruthy();
    });

    it('should render initial state of the component', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
  });

  describe('methods', () => {
    const emptyArray = [];

    describe('getNextAddressess', () => {
      it('should call vuex getNextWalletsFromHd with correct arguments', async () => {
        expect.assertions(3);

        wrapper.setMethods({
          getNextWalletsFromHd: jest.fn().mockResolvedValue([]),
        });

        await wrapper.vm.getNextAddressess({ offset: 10, limit: 20 });
        expect(wrapper.vm.addresses).toEqual(emptyArray);
        expect(wrapper.vm.getNextWalletsFromHd).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.getNextWalletsFromHd).toBeCalledWith({
          offset: 10,
          limit: 20,
          walletType: WALLET_TYPE.HD_PUBLIC,
        });
      });

      it('should call changePage with correct arguments', async () => {
        expect.assertions(3);

        wrapper.setMethods({
          getNextWalletsFromHd: jest.fn().mockResolvedValue([]),
        });

        await wrapper.vm.changePage(3);
        expect(wrapper.vm.addresses).toEqual(emptyArray);
        expect(wrapper.vm.getNextWalletsFromHd).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.getNextWalletsFromHd).toBeCalledWith({
          offset: 20,
          limit: 10,
          walletType: WALLET_TYPE.HD_PUBLIC,
        });
      });
    });

    describe('handleImport', () => {
      const password = 'password';

      beforeEach(() => {
        router.push('/kek');
        wrapper.setMethods({
          getNextWalletsFromHd: jest.fn().mockResolvedValue(addresses),
          addPublicWallet: jest.fn().mockResolvedValue(),
          addHdChildWallets: jest.fn().mockResolvedValue(),
        });
        wrapper.setData({
          activeAddress: address,
          addresses,
        });
      });

      it('should call vuex addPublicWallet with correct arguments and redirect to root', async () => {
        expect.assertions(6);

        wrapper.setComputed({
          isHDv3WalletByType: jest.fn().mockReturnValue(false),
        });

        expect(wrapper.vm.isPasswordModal).toBe(false);
        await wrapper.vm.handleImport();
        expect(wrapper.vm.isPasswordModal).toBe(false);

        expect(wrapper.vm.isHDv3WalletByType).toHaveBeenCalledTimes(2);
        expect(wrapper.vm.addPublicWallet).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.addPublicWallet).toBeCalledWith({
          address,
          info: {
            type: WALLET_TYPE.HD_PUBLIC,
            index: addressPos,
          },
        });

        expect(router.currentRoute.fullPath).toBe('/');
      });

      it('should call vuex addHdChildWallets with correct arguments and redirect to root', async () => {
        expect.assertions(6);

        wrapper.setComputed({
          isHDv3WalletByType: jest.fn().mockReturnValue(true),
        });
        router.push('/kek');

        expect(wrapper.vm.isPasswordModal).toBe(false);
        await wrapper.vm.handleImport();
        expect(wrapper.vm.isPasswordModal).toBe(true);
        await wrapper.vm.handleImportByPassword(password);

        expect(wrapper.vm.isHDv3WalletByType).toHaveBeenCalledTimes(2);
        expect(wrapper.vm.addHdChildWallets).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.addHdChildWallets).toBeCalledWith({
          address,
          index: addressPos,
          password,
          type: WALLET_TYPE.HD_PUBLIC,
        });
        expect(router.currentRoute.fullPath).toBe('/');
      });

      it('should call throw error', async () => {
        expect.assertions(4);

        const spy = jest.spyOn(wrapper.vm, '$notify');

        wrapper.setMethods({
          getNextWalletsFromHd: jest.fn().mockResolvedValue([]),
        });

        wrapper.setComputed({
          isHDv3WalletByType: jest.fn().mockReturnValue(false),
        });

        wrapper.setData({
          activeAddress: null,
        });

        router.push('/kek');

        await wrapper.vm.addWallet();

        expect(wrapper.vm.isHDv3WalletByType).toHaveBeenCalledTimes(0);

        expect(wrapper.vm.$notify).toHaveBeenCalledTimes(1);
        expect(wrapper.vm.$notify).toHaveBeenCalledWith({
          title: 'Importing error',
          text: 'An error occurred while importing wallet. Please try again.',
          type: 'is-danger',
        });

        expect(router.currentRoute.fullPath).toBe('/kek');

        spy.mockRestore();
      });
    });
  });
});
