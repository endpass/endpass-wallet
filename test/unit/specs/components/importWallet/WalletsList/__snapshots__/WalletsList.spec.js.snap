// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletsList render should render initial state of the component 1`] = `
<div
  class="column"
>
  <div
    class="buttons"
  >
    <v-button
      class-name="is-primary is-medium"
    >
      
      Load wallets
    
    </v-button>
  </div>
   
  <!---->
   
  <div
    style="display: none;"
  >
     
    <v-pagination />
     
    <div
      class="buttons"
    >
      <v-button
        class-name="is-primary is-medium"
        data-test="import-wallet-button"
        disabled="true"
      >
        
        Import
      
      </v-button>
    </div>
  </div>
   
  <!---->
</div>
`;
